### PR TITLE
[Integration][SonarQube] - Add filters to Projects and Issues

### PR DESCRIPTION
**Pull request by PeyGis**
# Description

Customers are unable to filter the parameters passed to the SonarQube API. Issue reported [here](https://github.com/port-labs/ocean/issues/419)
What -
Why -
How - Added apiQueryParams to each integration resource config, to allow clients to add necessary filters to ingest the desired data from their Sonar instance. Updated the config to ingest unresolved issues by default

## Type of change

Please leave one option from the following and delete the rest:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Screenshots

Project kind
Screenshot 2024-05-09 at 1 07 43 PM

Issue kind
Screenshot 2024-05-09 at 1 09 11 PM

## API Documentation

Details on the projects/components API query parameters can be found [here](https://next.sonarqube.com/sonarqube/web_api/api/components/search_projects?internal=true)

Details on the issues query parameters can be found [here](https://next.sonarqube.com/sonarqube/web_api/api/issues/search)

### CODE REVIEW

File: integrations/sonarqube/client.py [hunk 1]
```diff
@@ -97,6 +108,10 @@ async def send_paginated_api_request(
         json_data: Optional[dict[str, Any]] = None,
     ) -> list[dict[str, Any]]:
         try:
+            logger.info(
```
Comment (shalev007) [inline integrations/sonarqube/client.py:111]: This is great for debugging but I think it will probably spam the logs, maybe try to use debug

```suggestion
            logger.debug(
```

File: integrations/sonarqube/client.py [hunk 2]
```diff
@@ -256,17 +292,27 @@ async def get_issues_by_component(
...
-            query_params={"componentKeys": component_key},
+            query_params=query_params,
         )
 
         for issue in response:
-            issue["__link"] = (
-                f"{self.base_url}/project/issues?open={issue.get('key')}&id={component_key}"
-            )
+            issue[
+                "__link"
```
Comment (shalev007) [inline integrations/sonarqube/client.py:314]: I know this isn't new code but, I think it's best to use a constant for these types of custom properties, WDYT?

File: integrations/sonarqube/client.py [hunk 3]
```diff
@@ -256,17 +292,27 @@ async def get_issues_by_component(
...
+        query_params = {"componentKeys": component_key}
         logger.info(f"Fetching all issues in component: {component_key}")
+
+        selector = typing.cast(
+            SonarQubeIssueResourceConfig, event.resource_config
+        ).selector
+
+        if hasattr(selector, "api_query_params") and selector.api_query_params:
+            selector_query_params = selector.api_query_params.generate_request_params()
+            query_params.update(selector_query_params)
```
Comment (shalev007) [inline integrations/sonarqube/client.py:304]: could you make a function for this?
I see it's a duplicated in `get_components` and it will simplify the code

File: integrations/sonarqube/main.py [hunk 4]
```diff
@@ -28,14 +21,13 @@ def init_sonar_client() -> SonarQubeClient:
...
 async def on_project_resync(kind: str) -> list[dict[str, Any]]:
     logger.info(f"Listing Sonarqube resource: {kind}")
+
     sonar_client = init_sonar_client()
     return await sonar_client.get_all_projects()
 
 
 @ocean.on_resync(ObjectKind.ISSUES)
 async def on_issues_resync(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
-    logger.info(f"Listing Sonarqube resource: {kind}")
```
Comment (shalev007) [inline integrations/sonarqube/main.py:37]: Why did you remove all the logs?

File: integrations/sonarqube/integration.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        if resolved := value.pop("resolved", None):
+            value["resolved"] = resolved
+        if rules := value.pop("rules", None):
+            value["rules"] = ",".join(rules)
+        if scopes := value.pop("scopes", None):
+            value["scopes"] = ",".join(scopes)
+        if sonarsource_security := value.pop("sonarsource_security", None):
+            value["sonarsourceSecurity"] = ",".join(sonarsource_security)
+        if tags := value.pop("tags", None):
+            value["tags"] = ",".join(tags)
```
Comment (shalev007) [inline integrations/sonarqube/integration.py:138]: This looks a bit messy, is there a way to use `switch case` or something to make the code more readable?

File: integrations/sonarqube/main.py (hunk 4 above)
Comment (PeyGis) [reply integrations/sonarqube/main.py:37]: The ocean framework already logs these resync event for each kind. So adding another log makes it redundant

File: integrations/sonarqube/client.py (hunk 2 above)
Comment (PeyGis) [reply integrations/sonarqube/client.py:314]: that's great and something we'll need to discuss with Yair and Tom as we decided to distinguish custom properties from the one returned by the third party API using __<propName>

File: integrations/sonarqube/client.py [hunk 6]
```diff
@@ -132,22 +149,47 @@ async def send_paginated_api_request(
             logger.error(f"HTTP occurred while fetching paginated data: {e}")
             raise
 
+    def process_selector_query_params(self, selector: Selector) -> dict[str, Any]:
+        if hasattr(selector, "api_query_params") and selector.api_query_params:
+            return selector.api_query_params.generate_request_params()
+        return {}
```
Comment (yairsimantov20) [inline integrations/sonarqube/client.py:155]: Can u instead just override the resourceConfig implementation and inherit this functionality to exists in every of the kinds?

```
class MY_BASE_CUSTOM_SELECTOR(Selector):
    def generate_request_params(self):
        return {}
        
class MY_CUSTOM_SELECTOR(MY_BASE_CUSTOM_SELECTOR):
    ...

class MY_SONAR_CUSTOM_RESOURCE(ResourceConfig):
    selector: MY_CUSTOM_SELECTOR
        
class B(MY_SONAR_CUSTOM_RESOURCE):
  ...

File: integrations/sonarqube/.port/resources/port-app-config.yaml [hunk 7]
```diff
@@ -61,6 +61,8 @@ resources:
   - kind: issues
     selector:
       query: 'true'
+      apiQueryParams:
+        resolved: 'false'
```
Comment (yairsimantov20) [inline integrations/sonarqube/.port/resources/port-app-config.yaml:65]: can u also allow `false` & `true` without `'` in the values
i think itll be dummy proof

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
comments

File: integrations/sonarqube/integration.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                    else:
+                        params.append(f"{field}={value}")
+                else:
+                    params.append(f"{field}={value}")
+        return " and ".join(params)
+
+
+class SonarQubeProjectAPIQueryParams(BaseModel):
+    q: str | None
+    s: str | None
```
[the remark this answers is not in the document]
Comment (PeyGis) [reply integrations/sonarqube/integration.py:51]: Yes, that's the naming in the API. I will add the description

File: integrations/sonarqube/integration.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    ISSUES = "issues"
+    ANALYSIS = "analysis"
+    SASS_ANALYSIS = "saas_analysis"
+    ONPREM_ANALYSIS = "onprem_analysis"
+
+
+class SonarQubeComponentSearchFilter(BaseModel):
+    query: str | None
+    metrics: list[dict[str, str]] | None
+    alert_status: str | None
```
[the remark this answers is not in the document]
Comment (PeyGis) [reply integrations/sonarqube/integration.py:26]: I've added it

File: integrations/sonarqube/integration.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+                        for metric_key, metric_value in metric_filter.items():
+                            params.append(f"{metric_key} {metric_value}")
+                elif field in ["languages", "tags"]:
+                    if isinstance(value, list):
+                        params.append(f"{field} IN ({','.join(value)})")
+                    else:
+                        params.append(f"{field}={value}")
+                else:
+                    params.append(f"{field}={value}")
+        return " and ".join(params)
```
[the remark this answers is not in the document]
Comment (PeyGis) [reply integrations/sonarqube/integration.py:46]: Then the final result will not include the "and"

An example is shown below when you use only one of the params:

'filter': 'alert_status=ERROR'

File: integrations/sonarqube/.port/resources/port-app-config.yaml (hunk 7 above)
Comment (PeyGis) [reply integrations/sonarqube/.port/resources/port-app-config.yaml:65]: This will not work since Pydantic will convert it to a Python bool (True or False), and this is not one of the allowed values

File: integrations/sonarqube/client.py [hunk 11]
```diff
@@ -138,16 +153,38 @@ async def get_components(self) -> list[dict[str, Any]]:
...
+                f"Fetching all components in organization: {self.organization_id}"
+            )
+
+        app_config_resources = event.port_app_config.resources
+
+        project_resource = None
+        for resource in app_config_resources:
+            if resource.kind == ObjectKind.PROJECTS:
+                project_resource = resource
+                break
```
[the remark this answers is not in the document]
Comment (PeyGis) [reply integrations/sonarqube/client.py:176]: The intent is not to get the current resource but the project resource. The reason is that this method get's called in several places and different resync functions: analysis, issues etc

So if you get the current project, you will not get the expected behaviour. For instance, before I sync all issues, I want to fetch only the components that have been defined by the user in the config mapping, grab the filtering on the projects and sync issues for ONLY the filtered components.

File: integrations/sonarqube/integration.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+class SonarQubeProjectAPIQueryParams(BaseModel):
+    q: str | None
+    s: str | None
+    filter: SonarQubeComponentSearchFilter | None
+
+    def generate_request_params(self) -> dict[str, Any]:
+        value = self.dict(exclude_none=True)
+        if filter := value.pop("filter", None):
+            filter_instance = SonarQubeComponentSearchFilter(**filter)
+            value["filter"] = filter_instance.generate_search_filters()
```
[the remark this answers is not in the document]
Comment (PeyGis) [reply integrations/sonarqube/integration.py:71]: The `generate_search_filters` is already part of the BaseModel class (`SonarQubeComponentSearchFilter`)
Can you please rephrase as I didn't get what you mean clearly

File: integrations/sonarqube/client.py (hunk 2 above)
Comment (shalev007) [reply integrations/sonarqube/client.py]: The property name is great!
I only thought since your using it in a lot of places it would be best to save it in a constant for reusability

Comment (PeyGis) [reply integrations/sonarqube/client.py]: ahh got it now. Yeah, I'll make it constant

File: integrations/sonarqube/client.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (yairsimantov20) [reply integrations/sonarqube/client.py:169]: what if there are multiple project resources?

[the remark this answers is not in the document]
Comment (yairsimantov20) [reply integrations/sonarqube/client.py:169]: what if someone wants to query all components?

File: integrations/sonarqube/CHANGELOG.md [hunk 13]
```diff
@@ -7,6 +7,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 <!-- towncrier release notes start -->
 
+# Port_Ocean 0.1.44 (2024-04-18)
```
Comment (yairsimantov20) [inline integrations/sonarqube/CHANGELOG.md:10]:
```suggestion
# Port_Ocean 0.1.44 (2024-05-02)
```

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
last comment

File: integrations/sonarqube/integration.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    def generate_request_params(self) -> dict[str, Any]:
+        value = self.dict(exclude_none=True, by_alias=True)
+
+        for key in list(value.keys()):
+            val = value[key]
+            value[key] = ",".join(val) if isinstance(val, list) else val
+        return value
+
+
+class CustomSelector(Selector):
```
Comment (yairsimantov20) [inline integrations/sonarqube/integration.py:144]: make you classes inherit the same class that implement `api_query_params` with default return of `{}` and then just cast the selector to this type if needed

File: integrations/sonarqube/main.py [hunk 15]
```diff
@@ -26,16 +19,17 @@ def init_sonar_client() -> SonarQubeClient:
...
 @ocean.on_resync(ObjectKind.PROJECTS)
-async def on_project_resync(kind: str) -> list[dict[str, Any]]:
+async def on_project_resync(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
     logger.info(f"Listing Sonarqube resource: {kind}")
+
     sonar_client = init_sonar_client()
-    return await sonar_client.get_all_projects()
+    async for project_list in sonar_client.get_all_projects():
+        yield project_list
+    # return await sonar_client.get_all_projects()
```
Comment (yairsimantov20) [inline integrations/sonarqube/main.py:28]: ?

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
more comments

File: integrations/sonarqube/CHANGELOG.md [hunk 16]
```diff
@@ -6,6 +6,12 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 <!-- towncrier release notes start -->
+# Port_Ocean 0.1.48 (2024-05-07)
```
Comment (yairsimantov20) [inline integrations/sonarqube/CHANGELOG.md:9]:
```suggestion
# Port_Ocean 0.1.48 (2024-05-16)
```

**Review by yairsimantov20 — APPROVED** [review]
🌊
